src/System.IO/tests/Stream/Stream.CopyToTests.cs

```diff
@@ -0,0 +1,338 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public class StreamCopyToTests
+    {
+        [Fact]
+        public void IfCanSeekIsFalseLengthAndPositionShouldNotBeCalled()
+        {
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => false,
+                readFunc: (array, index, length) => 0); // (buffer, offset, count) declared in enclosing scope
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            trackingStream.CopyTo(dest);
+
+            Assert.True(trackingStream.TimesCalled(nameof(trackingStream.CanSeek)) <= 1);
+            Assert.Equal(0, trackingStream.TimesCalled(nameof(trackingStream.Length)));
+            Assert.Equal(0, trackingStream.TimesCalled(nameof(trackingStream.Position)));
+            // We can't override CopyTo since it's not virtual, so checking TimesCalled
+            // for CopyTo will result in 0. Instead, we check that Read was called,
+            // and validate the parameters passed there.
+            Assert.Equal(1, trackingStream.TimesCalled(nameof(trackingStream.Read)));
+
+            byte[] buffer = trackingStream.ReadBuffer;
+            int offset = trackingStream.ReadOffset;
+            int count = trackingStream.ReadCount;
+
+            Assert.NotNull(buffer);
+            Assert.True(offset >= 0 && offset + count <= buffer.Length);
+            Assert.True(count > 0); // the buffer can't be size 0
+        }
+
+        [Fact]
+        public async void AsyncIfCanSeekIsFalseLengthAndPositionShouldNotBeCalled()
+        {
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => false,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            await trackingStream.CopyToAsync(dest);
+
+            Assert.True(trackingStream.TimesCalled(nameof(trackingStream.CanSeek)) <= 1);
+            Assert.Equal(0, trackingStream.TimesCalled(nameof(trackingStream.Length)));
+            Assert.Equal(0, trackingStream.TimesCalled(nameof(trackingStream.Position)));
+            Assert.Equal(1, trackingStream.TimesCalled(nameof(trackingStream.CopyToAsync)));
+
+            Assert.Same(dest, trackingStream.CopyToAsyncDestination);
+            Assert.True(trackingStream.CopyToAsyncBufferSize > 0);
```

Comment:
Another example for InRange.  I'll stop commenting on these. You get the idea :)